File: extensions/defenseclaw/src/scanners/plugin_scanner/analyzer_factory.ts
/**
 * Centralized analyzer construction — modeled after the skill scanner's
 * build_analyzers() in analyzer_factory.py.
 *
 * Every entry point (CLI binary, VSCode extension, API) MUST build analyzers
 * through this function so that:
 * - All core analyzers are always included.
 * - Analyzer toggling via disabledAnalyzers is respected.
 * - LLM analyzers are included when configured.
 * - Adding or removing a core analyzer only requires a change here.
 */
import type { Analyzer, BuildAnalyzersOptions } from "./analyzer.js";
import type { LLMPolicy } from "./policy.js";
import {
  PermissionsAnalyzer,
  DependencyAnalyzer,
  InstallScriptAnalyzer,
  ToolAnalyzer,
  SourceAnalyzer,
  DirectoryStructureAnalyzer,
  ClawManifestAnalyzer,
  BundleSizeAnalyzer,
  JsonConfigAnalyzer,
  LockfileAnalyzer,
  MetaAnalyzer,
} from "./analyzer_classes.js";
import { LLMAnalyzer } from "./llm_analyzer.js";

/**
 * Build the full analyzer pipeline.
 *
 * Core analyzers are always included unless explicitly disabled.
 * LLM analyzer is included when llmPolicy.enabled is true.
 * MetaAnalyzer runs last with LLM enhancement when available.
 */
export function buildAnalyzers(options?: BuildAnalyzersOptions & { llm?: LLMPolicy }): Analyzer[] {
  const disabled = new Set(options?.disabledAnalyzers ?? []);

  // Phase 1: Pattern-based analyzers
  const core: Analyzer[] = [
    new PermissionsAnalyzer(),
    new DependencyAnalyzer(),
    new InstallScriptAnalyzer(),
    new ToolAnalyzer(),
    new SourceAnalyzer(),
    new DirectoryStructureAnalyzer(),
    new ClawManifestAnalyzer(),
    new BundleSizeAnalyzer(),
    new JsonConfigAnalyzer(),
    new LockfileAnalyzer(),
  ];

  const filtered = core.filter((a) => !disabled.has(a.name));

  // Phase 2: LLM analyzer (opt-in, runs after pattern analyzers)
  if (options?.llm?.enabled && !disabled.has("llm")) {
    filtered.push(new LLMAnalyzer({
      model: options.llm.model,
      apiKey: options.llm.api_key || undefined,
      apiBase: options.llm.api_base || undefined,
      provider: options.llm.provider || undefined,
      maxTokens: options.llm.max_output_tokens,
      consensusRuns: options.llm.consensus_runs,
      pythonBinary: options.llm.python_binary || undefined,
    }));
  }

  // Phase 3: Meta analyzer (last — cross-references all previous findings)
  // Gets LLM config for optional LLM-powered validation/correlation
  if (!disabled.has("meta")) {
    filtered.push(new MetaAnalyzer(options?.llm));
  }

  return filtered;
}

File: cli/defenseclaw/scanner/plugin_scanner/analyzer_factory.py
"""Centralized analyzer construction."""

from __future__ import annotations

from defenseclaw.scanner.plugin_scanner.analyzer_classes import (
    BundleSizeAnalyzer,
    ClawManifestAnalyzer,
    DependencyAnalyzer,
    DirectoryStructureAnalyzer,
    InstallScriptAnalyzer,
    JsonConfigAnalyzer,
    LockfileAnalyzer,
    MetaAnalyzer,
    PermissionsAnalyzer,
    SourceAnalyzer,
    ToolAnalyzer,
)


def build_analyzers(
    *,
    profile: str = "default",
    disabled_analyzers: list[str] | None = None,
    llm: dict | None = None,
) -> list:
    """Build the full analyzer pipeline.

    Core analyzers are always included unless explicitly disabled.
    LLM analyzer is included when llm["enabled"] is true.
    MetaAnalyzer runs last with LLM enhancement when available.
    """
    disabled = set(disabled_analyzers or [])

    # Phase 1: Pattern-based analyzers
    core = [
        PermissionsAnalyzer(),
        DependencyAnalyzer(),
        InstallScriptAnalyzer(),
        ToolAnalyzer(),
        SourceAnalyzer(),
        DirectoryStructureAnalyzer(),
        ClawManifestAnalyzer(),
        BundleSizeAnalyzer(),
        JsonConfigAnalyzer(),
        LockfileAnalyzer(),
    ]

    filtered = [a for a in core if a.name not in disabled]

    # Phase 2: LLM analyzer (opt-in, runs after pattern analyzers)
    if llm and llm.get("enabled") and "llm" not in disabled:
        from defenseclaw.scanner.plugin_scanner.llm_analyzer import LLMAnalyzer

        filtered.append(
            LLMAnalyzer(
                {
                    "model": llm.get("model", ""),
                    "api_key": llm.get("api_key") or None,
                    "api_base": llm.get("api_base") or None,
                    "provider": llm.get("provider") or None,
                    "max_tokens": llm.get("max_output_tokens"),
                    "consensus_runs": llm.get("consensus_runs"),
                    "python_binary": llm.get("python_binary") or None,
                }
            )
        )

    # Phase 3: Meta analyzer (last -- cross-references all previous findings)
    if "meta" not in disabled:
        filtered.append(MetaAnalyzer(llm))

    return filtered
